elastic/kibana
# [Upgrade Assistant] Update logic for handling reindex failures

Fixes https://github.com/elastic/kibana/issues/124153

The Upgrade Assistant was incorrectly marking a reindexing process as a failure. This PR adjusts the logic for determining a reindex failure to only check for the existence of `failures` in the task API response.

It also fixes a minor spacing issue when rendering an error callout.


### How to test

This is a hard one to add automated tests for (open to suggestions!), as it involves restarting Kibana to reproduce. 

I was able to reproduce the bug (pre change) fairly consistently using the manual steps documented in https://github.com/elastic/kibana/pull/123817#issue-1115001979. With this change, the reindexing process should no longer fail.

To trigger an actual reindexing failure (post change):

1. Start up 6.x Elasticsearch and create some indices that contain a property with a `text` type. For example:

```
PUT /test/_doc/1
{
  "field": "my_string"
}
```

Alternatively, you can use this zip file, which already contains some indices: [6.8.16-data.zip](https://github.com/elastic/kibana/files/8016668/6.8.16-data.zip)

2. Start Elasticsearch with the 6.x snapshot created in step 1: `yarn es snapshot --license=trial -E path.data=./path_to_6.x`
3. [Hack the UA code ](https://github.com/elastic/kibana/blob/7.17/x-pack/plugins/upgrade_assistant/server/lib/reindexing/reindex_service.ts#L196) so that it creates the destination index with a bad mapping. This will trigger a reindex failure:
```
      createIndex = await esClient.indices.create({
        index: newIndexName,
        body: {
          settings,
          mappings: {
            properties: {
              field: { type: 'date' },
            },
          },
        },
      });
```
4. Start Kibana and navigate to Stack Management -> Upgrade Assistant -> ES deprecations
5. Start reindexing one of the indices. Note reindexing should fail and trigger error callout.

### Screenshots
Before/after of spacing issue.

Before:
![Upgrade-Assistant-Elastic](https://user-images.githubusercontent.com/5226211/152407026-f27e7f69-f7b1-4e52-9ef2-23447a56ea15.png)

After:
<img width="552" alt="Screen Shot 2022-02-03 at 1 32 48 PM" src="https://user-images.githubusercontent.com/5226211/152407049-f511aede-9164-4c41-9347-b492a56f9327.png">
